expand functional map for reduced floating points on CPU

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105131
* #107128
* #104863
* #104693
* #103826
* __->__ #104584
* #104583

Return output in accumulated dtype in vec::reduce functions when input is float16 or bfloat16, to reduce rounding error.

cc @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

mingfeima committed:
expand functional map for reduced floating points on CPU

[ghstack-poisoned]

mingfeima committed:
Update on "expand functional map for reduced floating points on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "expand functional map for reduced floating points on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "expand functional map for reduced floating points on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "expand functional map for reduced floating points on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "expand functional map for reduced floating points on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "expand functional map for reduced floating points on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "expand functional map for reduced floating points on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "expand functional map for reduced floating points on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "expand functional map for reduced floating points on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "expand functional map for reduced floating points on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "expand functional map for reduced floating points on CPU"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "expand functional map for reduced floating points on CPU"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "expand functional map for reduced floating points on CPU"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "expand functional map for reduced floating points on CPU"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "expand functional map for reduced floating points on CPU"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "expand functional map for reduced floating points on CPU"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "expand functional map for reduced floating points on CPU"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "expand functional map for reduced floating points on CPU"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "expand functional map for reduced floating points on CPU"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "expand functional map for reduced floating points on CPU"


Return output in accumulated dtype in vec::reduce functions when input is float16 or bfloat16, to reduce rounding error.

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 reviewed (approved):
Overall looks good. Some suggestion on code comments fyi.

Valentine233 committed:
Update on "expand functional map for reduced floating points on CPU"


Return output in accumulated dtype in vec::reduce functions when input is float16 or bfloat16, to reduce rounding error.

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "expand functional map for reduced floating points on CPU"


Return output in accumulated dtype in vec::reduce functions when input is float16 or bfloat16, to reduce rounding error.

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima commented on aten/src/ATen/cpu/vec/functional_bfloat16.h:
```diff
@@ -404,6 +404,41 @@ inline void map(
+      data_fvec0 = fVec::loadu(input_data + d);
+      data_fvec1 = fVec::loadu(input_data + d + fVec::size(), size - d - fVec::size());
+    } else {
+      // choose to align with behaviour of bVec::louad(ptr, size),
```
oh thanks! @Valentine233 could you please address this change. I did not create a separate PR for this one since we don't have to make big changes.

Valentine233 replied:
Sure!

Valentine233 committed:
Update on "expand functional map for reduced floating points on CPU"


Return output in accumulated dtype in vec::reduce functions when input is float16 or bfloat16, to reduce rounding error.

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "expand functional map for reduced floating points on CPU"


Return output in accumulated dtype in vec::reduce functions when input is float16 or bfloat16, to reduce rounding error.

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 commented:
@pytorchbot merge

Valentine233 commented:
@pytorchbot merge
